### PR TITLE
ll_schedule: refine to schedule period task no later than half period

### PR DESCRIPTION
**Pull request by keyonjie**
For period task like DMIC/SSP pipeline, starting the first copying too
late may lead to xrun after the DAI is started.

Change to ask for earlier first-time scheduling (not later than half
period) to avoid this kind of issue as more as possible.

Signed-off-by: Keyon Jie <yang.jie@linux.intel.com>

### CODE REVIEW

**Review by lgirdwood — COMMENTED** [review]
Can you describe the problem this is solving ? Are you saying first copy() wrt to IPC trigger is too long ?

**Comment by keyonjie** [comment]
@lgirdwood this was created aimed to address the issue https://github.com/thesofproject/sof/issues/3939, but since the issue is very difficult to reproduce, let me make this as draft at the moment.

I will create another PR to add some useful information logs about the task scheduling point with ll_scheduler, which will help to capture logs once we hit the issue again.

File: src/schedule/ll_schedule.c [hunk 1]
```diff
@@ -285,7 +284,13 @@ static int schedule_ll_domain_set(struct ll_schedule_data *sch,
 		/* earlier periodic task, try to make it cadence-aligned with the existed task */
 		offset = (sch->domain->next_tick - task_start) %
 			 (sch->domain->ticks_per_ms * period / 1000);
-		task_start = task_start - task_start_ticks + offset;
+
+		/* start the period task not later than half period to avoid xrun */
+		deadline_delay_us = period >> 1;
+		if (offset > sch->domain->ticks_per_ms * deadline_delay_us)
+			offset -= sch->domain->ticks_per_ms * deadline_delay_us;
```
Comment (mwasko) [inline src/schedule/ll_schedule.c:291]: Based on description I think the intention was to set offset to half period when it is later than half period: offset = sch->domain->ticks_per_ms * deadline_delay_us; not current offset minus half period.

**Review by mwasko — CHANGES_REQUESTED** [review]
Please check my review comments.

**Comment by lgirdwood** [comment]
@keyonjie it sound like a race between IPC trigger(start) and the scheduler doing the work ?i.e. is IPC trggier start happens at the start of the scheduler tick then we have to wait upto 1ms for the next tick to do our work (but the DMA and DAI have started ??)

**Pull request closed without merging**